types/opt/value.go

```diff
@@ -0,0 +1,167 @@
+			// A nil pointer, unsafe pointer, or interface is equivalent to an unset value.
+			// It should be marshalled as null, or be omitted from the output.
+			return Empty[T]()
+		}
+	case reflect.Map, reflect.Slice:
+		if rv.IsNil() {
+			// A nil map or slice is equivalent to an empty map or slice,
+			// and should be marshalled as `{}` or `[]`, accordingly.
+			return Value[T]{value: v, set: true, isNil: true}
+		}
+	}
+	return Value[T]{value: v, set: true, isNil: false}
+}
+
+// String implements Stringer.
+func (o *Value[T]) String() string {
+	if !o.set {
+		return fmt.Sprintf("(empty[%T])", o.value)
+	}
+	return fmt.Sprint(o.value)
+}
+
+// Set assigns the specified value to the optional value o.
+func (o *Value[T]) Set(v T) {
+	*o = Of(v)
+}
+
+// Clear resets o to an empty state.
+func (o *Value[T]) Clear() {
+	*o = Empty[T]()
+}
+
+// Present reports whether o has a value set.
+func (o *Value[T]) Present() bool {
+	return o.set
+}
+
+// Get returns the value and a flag indicating whether the value is set.
+func (o *Value[T]) Get() (v T, ok bool) {
+	return o.value, o.set
+}
+
+// Equal reports whether o is equal to v.
+// Two optional values are equal if both are empty,
+// or if both are set and the underlying values are equal.
+// If the template type T implements an Equal(T) bool method, it will be used
+// instead of the == operator for value comparison.
+// If T is not comparable, it returns false.
+func (o *Value[T]) Equal(v *Value[T]) bool {
+	if o == nil && v == nil {
+		return true
+	}
+	if o == nil || v == nil || o.set != v.set {
+		return false
+	}
+	if !o.set {
+		return true
+	}
+	ov := any(o.value)
+	if eq, ok := ov.(equatable[T]); ok {
+		return eq.Equal(v.value)
+	}
+	if reflect.TypeFor[T]().Comparable() {
+		return ov == any(v.value)
+	}
+	return false
+}
+
+// EqualValue reports whether o is equal to v.
+// If the template type T implements an Equal(T) bool method, it will be used
+// instead of the == operator for value comparison.
+// If T is not comparable, it returns false.
+func (o *Value[T]) EqualValue(v T) bool {
+	if !o.set {
+		return false
+	}
+	ov := any(o.value)
+	if ov, ok := ov.(equatable[T]); ok {
+		return ov.Equal(v)
+	}
+	if reflect.TypeFor[T]().Comparable() {
+		return ov == any(v)
+	}
+	return false
+}
+
+// MarshalJSON implements json.Marshaler.
+func (o Value[T]) MarshalJSON() ([]byte, error) {
+	if !o.set {
+		return nullBytes, nil
+	}
+	if o.isNil {
+		// Marshal nil slices and nil maps as [] and {}, respectively.
+		switch reflect.TypeFor[T]().Kind() {
+		case reflect.Slice:
+			return nilSliceJSON, nil
+		case reflect.Map:
+			return nilMapJSON, nil
+		}
+	}
+	return json.Marshal(&o.value)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (o *Value[T]) UnmarshalJSON(j []byte) error {
+	if len(j) == 0 || bytes.Equal(j, nullBytes) {
+		*o = Empty[T]()
+		return nil
+	}
+	if err := json.Unmarshal(j, &o.value); err != nil {
+		return err
+	}
+	o.set, o.isNil = true, false
+	return nil
+}
```

Comment:
I wasn't sure if we're ready to move to the json-experiment encoder in more places. Additionally, this applies the new `omitempty` option behavior to `T`, so a zero value of the following struct results in the JSON output of `{"Int":0}` rather than `{}` as with the `encoding/json` encoder:
```
type testStruct struct {
	Int int    `json:",omitempty"`
	Str string `json:",omitempty"`
}
```

But I'm happy to use `json2`. Done.